### ISSUE TITLE
Utility functions?

### ISSUE THREAD

**Issue by pnrobinson** [monarch-initiative/N2V#105]
Should we add a file with various utility functions for working with the embeddings?
For instance, something like this to get the most similar words?
```
def get_cosine_sim(emb, valid_words, top_k):
    norm = np.sqrt(np.sum(emb**2,axis=1,keepdims=True))
    norm_emb = emb/norm
    in_emb = norm_emb[valid_words,:]
    similarity = np.dot(in_emb, np.transpose(norm_emb))
    sorted_ind = np.argsort(-similarity, axis=1)[:,1:top_k+1]
    return sorted_ind, valid_words
```

**Comment by callahantiff** [comment]
@pnrobinson - great idea! I'm happy to work on this. I know there are some embedding loading and writing functions in the scripts we could also pull out and add to a utility embedding script.

Will add a new branch for this off of `develop`.

**Comment by justaddcoffee** [comment]
+1 @callahantiff !

**Comment by vidarmehr** [comment]
@pnrobinson @callahantiff I think we may need adding Common Neighbors, Jaccard’s Coefficient, Adamic-Adar Score, Preferential Attachment scores, too. We don't need embedding to calculate these scores. But, I think we eventually need to calculate these scores.

**Comment by callahantiff** [comment]
> @pnrobinson @callahantiff I think we may need adding Common Neighbors, Jaccard’s Coefficient, Adamic-Adar Score, Preferential Attachment scores, too. We don't need embedding to calculate these scores. But, I think we eventually need to calculate these scores.

@vidarmehr, great point! Maybe we create a utility function for link prediction methods as well then?

**Comment by vidarmehr** [comment]
> > @pnrobinson @callahantiff I think we may need adding Common Neighbors, Jaccard’s Coefficient, Adamic-Adar Score, Preferential Attachment scores, too. We don't need embedding to calculate these scores. But, I think we eventually need to calculate these scores.
> 
> @vidarmehr, great point! Maybe we create a utility function for link prediction methods as well then?

Yes. I think that is a good idea to have a separate utility function for link prediction methods.

**Comment by pnrobinson** [comment]
Awesome! We should make these functions so that they are easily testable!

**Comment by callahantiff** [comment]
> Awesome! We should make these functions so that they are easily testable!

Absolutely! I will also open a second ticket for link prediction function metric utils.

**Comment by callahantiff** [comment]
@pnrobinson - perhaps we just want to use the tf built in cosine similarity function?

https://www.tensorflow.org/api_docs/python/tf/keras/metrics/CosineSimilarity

**Comment by vidarmehr** [comment]
> @pnrobinson - perhaps we just want to use the tf built in cosine similarity function?
> 
> https://www.tensorflow.org/api_docs/python/tf/keras/metrics/CosineSimilarity

I wrote a simple code to calculate cosine similarity. But, maybe it is good to use the tf implementation.

**Comment by pnrobinson** [comment]
As a rule it is almost always better to use library functions for things like this.

**Comment by pnrobinson** [comment]
@vidarmehr can we close this now?

**Comment by vidarmehr** [comment]
> @vidarmehr can we close this now?

Yes, Peter. I close it.
